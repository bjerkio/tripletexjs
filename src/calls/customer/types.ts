import { DefaultTripletexInputs } from '../../types';

export interface ListCustomersInput extends DefaultTripletexInputs {
  /**
   * List of IDs
   */
  id?: string[];

  /**
   * List of IDs
   */
  customerAccountNumber?: string[];

  /**
   * Equals
   */
  organizationNumber?: string;

  /**
   * Equals
   */
  email?: string;

  /**
   * Equals
   */
  invoiceEmail?: string;

  /**
   * Equals
   *
   * @default false
   */
  isInactive?: boolean;

  /**
   * List of IDs
   */
  accountManagerId?: string[];

  /**
   * Only return elements that have changed since this date and time
   */
  changedSince?: Date;
}
