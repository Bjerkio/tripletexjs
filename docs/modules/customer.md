[tripletexjs](../README.md) › [Customer](customer.md)

# Namespace: Customer

## Index

### Enumerations

* [emailAttachmentType](../enums/customer.emailattachmenttype.md)
* [invoiceSendMethod](../enums/customer.invoicesendmethod.md)
* [invoicesDueInType](../enums/customer.invoicesdueintype.md)
* [language](../enums/customer.language.md)

### Properties

* [accountManager](customer.md#optional-accountmanager)
* [bankAccounts](customer.md#optional-bankaccounts)
* [category1](customer.md#optional-category1)
* [category2](customer.md#optional-category2)
* [category3](customer.md#optional-category3)
* [changes](customer.md#optional-readonly-changes)
* [customerNumber](customer.md#optional-customernumber)
* [deliveryAddress](customer.md#optional-deliveryaddress)
* [description](customer.md#optional-description)
* [email](customer.md#optional-email)
* [id](customer.md#optional-id)
* [invoiceEmail](customer.md#optional-invoiceemail)
* [invoicesDueIn](customer.md#optional-invoicesduein)
* [isCustomer](customer.md#optional-readonly-iscustomer)
* [isInactive](customer.md#optional-readonly-isinactive)
* [isPrivateIndividual](customer.md#optional-isprivateindividual)
* [isSupplier](customer.md#optional-issupplier)
* [name](customer.md#name)
* [organizationNumber](customer.md#optional-organizationnumber)
* [overdueNoticeEmail](customer.md#optional-overduenoticeemail)
* [phoneNumber](customer.md#optional-phonenumber)
* [phoneNumberMobile](customer.md#optional-phonenumbermobile)
* [physicalAddress](customer.md#optional-physicaladdress)
* [postalAddress](customer.md#optional-postaladdress)
* [singleCustomerInvoice](customer.md#optional-singlecustomerinvoice)
* [supplierNumber](customer.md#optional-suppliernumber)
* [url](customer.md#optional-readonly-url)
* [version](customer.md#optional-version)

## Properties

### `Optional` accountManager

• **accountManager**? : *[Employee](employee.md)*

___

### `Optional` bankAccounts

• **bankAccounts**? : *Array‹string›*

List of the bank account numbers for this customer. Norwegian bank account numbers only.

___

### `Optional` category1

• **category1**? : *[CustomerCategory](../interfaces/customercategory.md)*

Category 1 of this customer

___

### `Optional` category2

• **category2**? : *[CustomerCategory](../interfaces/customercategory.md)*

Category 2 of this customer

___

### `Optional` category3

• **category3**? : *[CustomerCategory](../interfaces/customercategory.md)*

Category 3 of this customer

___

### `Optional` `Readonly` changes

• **changes**? : *Array‹[Change](change.md)›*

___

### `Optional` customerNumber

• **customerNumber**? : *undefined | number*

___

### `Optional` deliveryAddress

• **deliveryAddress**? : *[DeliveryAddress](../interfaces/deliveryaddress.md)*

___

### `Optional` description

• **description**? : *undefined | string*

___

### `Optional` email

• **email**? : *undefined | string*

___

### `Optional` id

• **id**? : *undefined | number*

___

### `Optional` invoiceEmail

• **invoiceEmail**? : *undefined | string*

___

### `Optional` invoicesDueIn

• **invoicesDueIn**? : *undefined | number*

Number of days/months in which invoices created from this customer is due

___

### `Optional` `Readonly` isCustomer

• **isCustomer**? : *undefined | false | true*

___

### `Optional` `Readonly` isInactive

• **isInactive**? : *undefined | false | true*

___

### `Optional` isPrivateIndividual

• **isPrivateIndividual**? : *undefined | false | true*

___

### `Optional` isSupplier

• **isSupplier**? : *undefined | false | true*

Defines if the customer is also a supplier.

___

###  name

• **name**: *string*

___

### `Optional` organizationNumber

• **organizationNumber**? : *undefined | string*

___

### `Optional` overdueNoticeEmail

• **overdueNoticeEmail**? : *undefined | string*

The email address of the customer where the noticing emails are sent in case of an overdue

___

### `Optional` phoneNumber

• **phoneNumber**? : *undefined | string*

___

### `Optional` phoneNumberMobile

• **phoneNumberMobile**? : *undefined | string*

___

### `Optional` physicalAddress

• **physicalAddress**? : *[Address](../interfaces/address.md)*

___

### `Optional` postalAddress

• **postalAddress**? : *[Address](../interfaces/address.md)*

___

### `Optional` singleCustomerInvoice

• **singleCustomerInvoice**? : *undefined | false | true*

Enables various orders on one customer invoice.

___

### `Optional` supplierNumber

• **supplierNumber**? : *undefined | number*

___

### `Optional` `Readonly` url

• **url**? : *undefined | string*

___

### `Optional` version

• **version**? : *undefined | number*
