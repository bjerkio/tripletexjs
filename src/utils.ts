import * as rt from 'runtypes';

export function pickFromObject<
  T extends Record<string, unknown>,
  K extends keyof T,
>(subject: T, ...keys: K[]): Pick<T, K> {
  const pairs = keys
    .map(key => [key, subject[key]])
    .filter(([, val]) => val !== undefined)
    .map(([key, val]) => [key, val]);
  return Object.fromEntries(pairs);
}

export function withRuntype<T>(validator: rt.Runtype<T>) {
  return (data: unknown) => {
    return validator.check(data);
  };
}

export function multipleValuesEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      from: rt.Number,
      count: rt.Number,
      versionDigest: rt.String.nullable().optional(),
      values: rt.Array(type).asReadonly(),
    })
    .asReadonly();
}

export function singleValueEnvelope<T>(type: rt.Runtype<T>) {
  return rt
    .Record({
      value: type,
    })
    .asReadonly();
}
