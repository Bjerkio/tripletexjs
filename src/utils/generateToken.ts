import { OpenAPI } from '../generated/core/OpenAPI';
import { TokenSessionService } from '../generated/services/TokenSessionService';
import { invariant } from './invariant';

/**
 * generateToken helps creating a token. It sets token for all requests.
 * 
 * @param consumerToken Token of the API consumer
 * @param employeeToken The employees token
 * @param expirationDate Expiration date for the combined token
 * @example
 * ```typescript
 * const expDate = new Date();
 * expDate.setDate(expDate.getDate() + 1);
 * await generateToken({
 *   consumerToken: '000-000-000',
 *   employeeToken: '100-100-100',
 *   expirationDate: expDate,
 * });
 * ```
 */
export async function generateToken({
  expirationDate,
  ...tokenSessionArgs
}: {
  consumerToken: string;
  employeeToken: string;
  expirationDate: Date;
}) {
  if (OpenAPI.TOKEN) {
    return false;
  }

  const expDate = expirationDate.toISOString().split('T')[0];

  const session = await TokenSessionService.tokenSessionCreateCreate({
    expirationDate: expDate,
    ...tokenSessionArgs,
  });

  invariant(session.value.token, 'Not able to get token');

  OpenAPI.TOKEN = session.value;
}

/**
 * 0 or blank means the company of the employee.
 * Any other value means accountant clients. Use /company/>withLoginAccess to get a list of those.
 * @param username defaults to '0'
 */
export async function setUsername(username: string | (() => Promise<string>)) {
  OpenAPI.USERNAME = username;
}