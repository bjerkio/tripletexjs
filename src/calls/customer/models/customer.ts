import * as rt from 'runtypes';
import { multipleValuesEnvelope, resourceRef } from '../../../utils';

// const addressRt = rt.Record({
//   id: rt.Number,
//   version: rt.Number,
//   employee: resourceRef,
//   addressLine1: rt.String,
//   addressLine2: rt.String,
//   postalCode: rt.String,
//   city: rt.String,
//   country: resourceRef,
//   name: rt.String.optional(),
// });

const customerCategoryRt = rt.Record({
  id: rt.Number,
  version: rt.Number,
  name: rt.String,
  number: rt.String,
  description: rt.String,
  type: rt.Number,
});

const companyBankAccountPresentationRt = rt.Record({
  iban: rt.String.nullable().optional(),
  bban: rt.String.nullable().optional(),
  bic: rt.String.nullable().optional(),
  country: resourceRef,
  provider: rt
    .Union(rt.Literal('NETS'), rt.Literal('AUTOPAY'))
    .nullable()
    .optional(),
});

const customerRt = rt.Record({
  id: rt.Number,
  version: rt.Number.nullable().optional(),
  // changes: rt.Array(changeRt),
  name: rt.String,
  organizationNumber: rt.String.nullable().optional(),
  supplierNumber: rt.Number.nullable().optional(),
  customerNumber: rt.Number.nullable().optional(),
  isSupplier: rt.Boolean,
  accountManager: resourceRef,
  email: rt.String.nullable().optional(),
  invoiceEmail: rt.String.nullable().optional(),
  overdueNoticeEmail: rt.String.nullable().optional(),
  bankAccounts: rt.Array(rt.String).nullable().optional(),
  phoneNumber: rt.String.nullable().optional(),
  phoneNumberMobile: rt.String.nullable().optional(),
  description: rt.String.nullable().optional(),
  language: rt.Union(rt.Literal('NO'), rt.Literal('EN'), rt.Literal('SV')),
  isPrivateIndividual: rt.Boolean.nullable().optional(),
  singleCustomerInvoice: rt.Boolean.nullable().optional(),
  invoiceSendMethod: rt.Union(
    rt.Literal('EMAIL'),
    rt.Literal('EHF'),
    rt.Literal('EFAKTURA'),
    rt.Literal('VIPPS'),
    rt.Literal('PAPER'),
    rt.Literal('MANUAL'),
  ),
  emailAttachmentType: rt.Union(rt.Literal('LINK'), rt.Literal('ATTACHMENT')),
  postalAddress: resourceRef,
  physicalAddress: resourceRef,
  deliveryAddress: resourceRef,
  category1: customerCategoryRt.nullable().optional(),
  category2: customerCategoryRt.nullable().optional(),
  category3: customerCategoryRt.nullable().optional(),
  invoicesDueIn: rt.Number.nullable().optional(),
  invoicesDueInType: rt.Union(
    rt.Literal('DAYS'),
    rt.Literal('MONTHS'),
    rt.Literal('RECURRING_DAY_OF_MONTH'),
  ),
  currency: resourceRef.nullable().optional(),
  bankAccountPresentation: rt.Array(companyBankAccountPresentationRt),
  isCustomer: rt.Boolean,
  isInactive: rt.Boolean,
});

export type Customer = rt.Static<typeof customerRt>;

export const listCustomerResponseRt = multipleValuesEnvelope(customerRt);
