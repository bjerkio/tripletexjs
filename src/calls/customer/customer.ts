import { serializeQuery, withRuntype } from '../../utils';
import { TripletexBase } from '../base';
import { listCustomerResponseRt } from './models/customer';
import { ListCustomersInput } from './types';
export * from './types';
export * from './models/customer';

export class TripletexCustomer extends TripletexBase {
  list(input?: ListCustomersInput) {
    const call = this.authenticatedCall() //
      .args<{
        input?: ListCustomersInput;
      }>()
      .path('/v2/customer')
      .query(args => (args.input ? serializeQuery(args.input) : {}))
      .method('get')
      .parseJson(withRuntype(listCustomerResponseRt))
      .build();

    return this.performRequest(sessionToken => call({ input, sessionToken }));
  }
}
