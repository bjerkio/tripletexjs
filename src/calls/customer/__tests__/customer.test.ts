import nock from 'nock';
import { invariant } from 'ts-invariant';
import { parseRuntypeValidationError } from '../../../utils';
import { TripletexCustomer } from '../customer';

const baseUrl = process.env.BASE_URL || 'https://api.tripletex.io/';
const sessionToken = process.env.SESSION_TOKEN || 'a-session-token';

describe('activity class', () => {
  let client: TripletexCustomer;

  beforeEach(() => {
    client = new TripletexCustomer({ baseUrl }, sessionToken);
  });

  it('should list activities', async () => {
    nock(baseUrl, { encodedQueryParams: true })
      .get('/v2/customer')
      .reply(200, {
        fullResultSize: 1,
        from: 0,
        count: 1,
        versionDigest: "'If-None-Match' header not specified",
        values: [
          {
            id: 11289298,
            version: 1,
            url: 'api.tripletex.io/v2/customer/11289298',
            name: 'Banana Airlines AB',
            organizationNumber: '',
            supplierNumber: 0,
            customerNumber: 10001,
            isSupplier: false,
            isCustomer: true,
            isInactive: false,
            accountManager: {
              id: 1734852,
              url: 'api.tripletex.io/v2/employee/1734852',
            },
            email: '',
            invoiceEmail: '',
            overdueNoticeEmail: '',
            bankAccounts: [],
            phoneNumber: '',
            phoneNumberMobile: '',
            description: '',
            language: 'NO',
            displayName: 'Banana Airlines AB (10001)',
            isPrivateIndividual: false,
            singleCustomerInvoice: false,
            invoiceSendMethod: 'EMAIL',
            emailAttachmentType: 'LINK',
            postalAddress: {
              id: 26087367,
              url: 'api.tripletex.io/v2/address/26087367',
            },
            physicalAddress: {
              id: 26087368,
              url: 'api.tripletex.io/v2/address/26087368',
            },
            deliveryAddress: null,
            category1: null,
            category2: null,
            category3: null,
            invoicesDueIn: 14,
            invoicesDueInType: 'DAYS',
            currency: { id: 1, url: 'api.tripletex.io/v2/currency/1' },
            bankAccountPresentation: [],
            ledgerAccount: {
              id: 31925215,
              url: 'api.tripletex.io/v2/ledger/account/31925215',
            },
            isFactoring: false,
          },
        ],
      });
    const activities = await client.list({});

    parseRuntypeValidationError(activities.error);

    invariant(activities.success);

    expect(activities.body.values).toMatchSnapshot();
  });
});
