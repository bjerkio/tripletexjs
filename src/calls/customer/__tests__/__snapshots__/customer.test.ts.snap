// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity class should list activities 1`] = `
Array [
  Object {
    "accountManager": Object {
      "id": 1734852,
      "url": "api.tripletex.io/v2/employee/1734852",
    },
    "bankAccountPresentation": Array [],
    "bankAccounts": Array [],
    "category1": null,
    "category2": null,
    "category3": null,
    "currency": Object {
      "id": 1,
      "url": "api.tripletex.io/v2/currency/1",
    },
    "customerNumber": 10001,
    "deliveryAddress": null,
    "description": "",
    "displayName": "Banana Airlines AB (10001)",
    "email": "",
    "emailAttachmentType": "LINK",
    "id": 11289298,
    "invoiceEmail": "",
    "invoiceSendMethod": "EMAIL",
    "invoicesDueIn": 14,
    "invoicesDueInType": "DAYS",
    "isCustomer": true,
    "isFactoring": false,
    "isInactive": false,
    "isPrivateIndividual": false,
    "isSupplier": false,
    "language": "NO",
    "ledgerAccount": Object {
      "id": 31925215,
      "url": "api.tripletex.io/v2/ledger/account/31925215",
    },
    "name": "Banana Airlines AB",
    "organizationNumber": "",
    "overdueNoticeEmail": "",
    "phoneNumber": "",
    "phoneNumberMobile": "",
    "physicalAddress": Object {
      "id": 26087368,
      "url": "api.tripletex.io/v2/address/26087368",
    },
    "postalAddress": Object {
      "id": 26087367,
      "url": "api.tripletex.io/v2/address/26087367",
    },
    "singleCustomerInvoice": false,
    "supplierNumber": 0,
    "url": "api.tripletex.io/v2/customer/11289298",
    "version": 1,
  },
]
`;
